### ISSUE TITLE
Error running deployment script

### ISSUE THREAD

**Issue by FunkyCat123** [OfficeDev/microsoft-teams-apps-callrecord-insights#20]
Tried this for two different subscriptions and getting the same error. I am an owner for both subscriptions.
Any help would be greatly appreciated.

Here is the error message:

deployifneeded : Failed to create Solution in resource group 'TeamsCDR'. Please ensure you have access to the subscription and try again.
At C:\temp\microsoft-teams-apps-callrecord-insights-main\deploy\deploy.ps1:402 char:15
+ $Deployment = deployifneeded 'Solution' deploy (@{
+               ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (:) [Write-Error], WriteErrorException
    + FullyQualifiedErrorId : Microsoft.PowerShell.Commands.WriteErrorException,deployifneeded
 
deployifneeded : ERROR: {"error":{"code":"InvalidTemplate","message":"Deployment template validation failed: 'The provided value for the template 
parameter 'graphChangeTrackingSPNObjectId' is not valid. Length of the value should be greater than or equal to '36'. Please see 
https://aka.ms/arm-syntax-parameters for usage details.'.","additionalInfo":[{"type":"TemplateViolation","info":{"lineNumber":98,"linePosition":21,"path"
:"properties.template.parameters.graphChangeTrackingSPNObjectId.minLength"}}]}}
At C:\temp\microsoft-teams-apps-callrecord-insights-main\deploy\deploy.ps1:402 char:15
+ $Deployment = deployifneeded 'Solution' deploy (@{
+               ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (ERROR: {"error"...minLength"}}]}}:String) [Write-Error], RemoteException
    + FullyQualifiedErrorId : NativeCommandErrorMessage,deployifneeded

**Comment by StefanFried** [comment]
i guess the change tracking app does not exist in your tenant. There is an article how you can create it if it does not exist
https://github.com/OfficeDev/microsoft-teams-apps-callrecord-insights/blob/main/docs/troubleshooting.md 
hope this helps :)

**Comment by FunkyCat123** [comment]
I was able to address this issue by changing line #128 of deploy.ps1 script as following:
getspnobjectid            = { az ad sp list --spn $args[0] --query "[].id" 2>&1 }
to 
getspnobjectid            = { az ad sp list --spn $args[0] --query "[].appId" 2>&1 }

Deploy.ps1 script is able to validate tracking app and continue but still fails few minutes later with the following error:

Configuring Function App Identity with required permissions.
Adding app role 'df021288-bdef-4463-88db-98f22de89214' to Service Principal '38c66d9a-89e0-4efe-bca5-10c039420294'.
C:\temp\microsoft-teams-apps-callrecord-insights-main\deploy\deploy.ps1 : Failed to add app role 'df021288-bdef-4463-88db-98f22de89214' to Service 
Principal '38c66d9a-89e0-4efe-bca5-10c039420294'. Please ensure you have access to the tenant and try again.
At line:7 char:1
+ .\deploy.ps1 @DeploymentParameters
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (:) [Write-Error], WriteErrorException
    + FullyQualifiedErrorId : Microsoft.PowerShell.Commands.WriteErrorException,deploy.ps1
 
C:\temp\microsoft-teams-apps-callrecord-insights-main\deploy\deploy.ps1 : ERROR: Not 
Found({"error":{"code":"Request_ResourceNotFound","message":"Resource '00000003-0000-0000-c000-000000000000' does not exist or one of its queried 
reference-property objects are not present.","innerError":{"date":"2024-04-22T12:53:14","request-id":"3401a021-3b9a-4da4-871e-4a0563238510","client-reque
st-id":"3401a021-3b9a-4da4-871e-4a0563238510"}}})
At line:7 char:1
+ .\deploy.ps1 @DeploymentParameters
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (ERROR: Not Foun...0563238510"}}}):String) [Write-Error], RemoteException
    + FullyQualifiedErrorId : NativeCommandError,deploy.ps1

**Comment by adthom** [comment]
So your fix did not resolve the issue, it just caused it to pass the appId instead of a valid objectId. @StefanFried's response is what appears to be happening here. Can you validate that the service principal exists in your tenant?

**Comment by FunkyCat123** [comment]
The object does exist. Here is what I observed running some of the "get" commands:

az ad sp list --display-name "Microsoft Graph Change Tracking" --output table

True              Microsoft Graph Change Tracking  0bf30f3b-4a52-48df-9a82-234910c4a086  f8cdef31-a31e-4b4a-93e4-5f571e91255a  False                      
  Microsoft Graph Change Tracking  aaa86172-2549-40f8-a9e3-dfac6493b578  ServicePrincipal  Microsoft.DirectoryServices.ServicePrincipal  Microsoft Service
s  Application             AzureADMultipleOrgs

az ad sp list --spn 0bf30f3b-4a52-48df-9a82-234910c4a086 --query "[].appId" 2>&1
[
  "0bf30f3b-4a52-48df-9a82-234910c4a086"
]

This is the code used in deploy.ps1 script:
az ad sp list --spn 0bf30f3b-4a52-48df-9a82-234910c4a086 --query "[].id" 2>&1
[]

Looking at full output for: 
az ad sp list --spn 0bf30f3b-4a52-48df-9a82-234910c4a086
There is no "id" parameter at all. Only appId parameter which equals to the same value as specified in the deploy.ps1 script.

Also running the deploy,ps1 script actually returns positive result for the tracking app:

$DeploymentParameters = @{
  ResourceGroupName = 'CommServicesGroup'
  #BaseResourceName  = 'cridemo1'
  SubscriptionId    = '1da91e88-9947-4b66-8888-646ac8549b6e'
  TenantDomain      = 'office365users.com'
}
.\deploy.ps1 @DeploymentParameters
Ensuring we are connected to subscription '1da91e88-9947-4b66-8888-646ac8549b6e'.
Connected to subscription '1da91e88-9947-4b66-8888-646ac8549b6e' in tenant '17887803-adc8-4605-8f56-d14cf5462830' as 'aberin@office365users.com'.

Getting SPN object id for the Microsoft Graph Change Tracking app.
SPN object id for the Microsoft Graph Change Tracking app is '0bf30f3b-4a52-48df-9a82-234910c4a086'.

Curious what is "38c66d9a-89e0-4efe-bca5-10c039420294" in 
Failed to add app role 'df021288-bdef-4463-88db-98f22de89214' to Service 
Principal '38c66d9a-89e0-4efe-bca5-10c039420294'. Please ensure you have access to the tenant and try again.

**Comment by FunkyCat123** [comment]
> df021288-bdef-4463-88db-98f22de89214

Thank you. This was the right direction. After changing appId parameter to objectId in the query, it allowed to move forward. At this point stuck at 
Triggering function to add the Call Records Graph Subscription (https://commservicesgroup-function.azurewebsites.net/api/subscription).

The script keeps retrying. Holding my fingers crossed :)

**Comment by FunkyCat123** [comment]
> > df021288-bdef-4463-88db-98f22de89214
> 
> Thank you. This was the right direction. After changing appId parameter to objectId in the query, it allowed to move forward. At this point stuck at Triggering function to add the Call Records Graph Subscription (https://commservicesgroup-function.azurewebsites.net/api/subscription).
> 
> The script keeps retrying. Holding my fingers crossed :)

Looks like the authorization is not going through even after an hour wait:

Failed to trigger function to add the Call Records Graph Subscription. The application is running but is not yet authorized to add the Call Records Graph 
Subscription. Try again later.

**Comment by adthom** [comment]
I have seen the permissions sometimes take a few hours to propagate. It may just be a matter of waiting here.
Can you confirm that Teams has been used in this tenant before? If no-one has ever signed into Teams in this tenant, then the subscription will never provision. 
Also, were you seeing the response from the spn query show an `objectId` property and not have an `id` property? Can you share the version of Azure CLI you are using  (`az --version`)?

**Comment by adthom** [comment]
Can you see if you update to Azure CLI 2.56+ if you see the same behavior? (See: [Requirements](https://github.com/OfficeDev/microsoft-teams-apps-callrecord-insights/wiki/Requirements#deployment-client-requirements)) That service principal object should just be coming from Graph API, which currently shows that field as being `id` ([ref](https://learn.microsoft.com/en-us/graph/api/resources/serviceprincipal?view=graph-rest-1.0#properties))

Were you able to verify that this tenant has used Teams previously?

**Comment by FunkyCat123** [comment]
After 24 hours still stuck with "Application is not yet authorized to add the Call Records Graph Subscription, awaiting app role propagation"

**Comment by adthom** [comment]
I think at this point, permissions have been applied, but perhaps the SPN challenges before have prevented the correct permissions from being applied to the Graph Key Vault secret. The `GraphEventHubConnectionString` secret (just this specific secret) in the Key Vault needs to have the [Key Vault Secrets User role](https://docs.microsoft.com/azure/role-based-access-control/built-in-roles#key-vault-secrets-user) granted to the `Microsoft Graph Change Tracking` service principal. Due to the issue in your deployment, I believe this is not currently assigned.

**Comment by StefanFried** [comment]
i see the same message across different tenants ("Application is not yet authorized to add the Call Records Graph Subscription, awaiting app role propagation") as funkycat123 does. The graph tracking spn has the correct role on the keyvault. So i can't explain the reason for this message because the deployment just runs fine until that point.'

If i call the url /api/subscription with they master key via postman i get a http 500 back
Interestingly i have a one tenant where the deployment just worked fine :)

**Comment by FunkyCat123** [comment]
Thank you all guys for your help!
When I checked key vault permissions it included an unknown entry which is most likely due to me messing with the deploy script earlier. I tried running the script for another resource group, but that didn't work. Probably because of everything that has been already created.
After looking at my bill for one day of use, I decided to stop the deployment. I was hoping to use this as a solution for customers that require retention of more than 28 days for CDR. However, the cost is way too much considering there is no data.
